### PR TITLE
AS-835: Integration test for deleting BQ dataset upon deleting a workspace

### PR DESCRIPTION
**Pull request by ansingh7115**
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/SnapshotAPISpec.scala [hunk 1]
```diff
@@ -199,6 +199,44 @@ class SnapshotAPISpec extends AnyFreeSpecLike with Matchers with BeforeAndAfterA
       }
     }
 
+    "should delete BigQuery dataset when a snapshot reference is deleted" taggedAs(Tags.AlphaTest, Tags.ExcludeInFiab) in {
```
Comment (davidangb) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/SnapshotAPISpec.scala]: fyi this isn't the lifecycle - we only delete the companion when the *workspace* is deleted, not when a reference is deleted

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/SnapshotAPISpec.scala [hunk 2]
```diff
@@ -314,6 +355,11 @@ class SnapshotAPISpec extends AnyFreeSpecLike with Matchers with BeforeAndAfterA
       content = payload)
   }
 
+  private def deleteSnapshotReference(projectName: String, workspaceName: String, resourceId: String)(implicit authToken: AuthToken) = {
+    val targetRawlsUrl  = Uri(Rawls.url).withPath(Path(s"/api/workspaces/$projectName/$workspaceName/snapshots/v2/$resourceId"))
+    Rawls.deleteRequest(uri = targetRawlsUrl.toString())
+  }
```
Comment (ansingh7115) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/SnapshotAPISpec.scala:363]: This actually ended up not being used but I would advocate for keeping it here since future tests will probably want to test deleting a snapshot reference.

**Review by MatthewBemis — APPROVED** [review]

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/SnapshotAPISpec.scala [hunk 3]
```diff
@@ -199,6 +199,47 @@ class SnapshotAPISpec extends AnyFreeSpecLike with Matchers with BeforeAndAfterA
...
+          createSnapshotReference(projectName, workspaceName, dataRepoSnapshotId, snapshotName)(owner.makeAuthToken())
+
+          // validate the snapshot was added correctly: list snapshots in Rawls, should return 1, which we just added.
+          val listResponse = listSnapshotReferences(projectName, workspaceName)
+          val resources = Rawls.parseResponseAs[ResourceList](listResponse).getResources.asScala
+          resources should have size 1
+
+          // check that the bq dataset has been created
+          val workspaceId = resources.head.getMetadata.getWorkspaceId
+          val datasetName = "deltalayer_forworkspace_" + workspaceId.toString.replace('-', '_')
```
Comment (ungwudik) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/SnapshotAPISpec.scala]: Nit: Looks like `deltalayer_forworkspace_` is the generic prefix for all datasets created by Delta Layer? Can it be declared as a constant with a line of comment that notes this prefix expectation?
Also, looks like there's an existing test in this class that can reuse the constant. i.e. `should add a BigQuery dataset for each snapshot reference added`

Comment (davidangb) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/SnapshotAPISpec.scala]: what you really want is `DeltaLayer.generateDatasetNameForWorkspace()` … but that's in the main Rawls codebase, not automation. I realize this comment doesn't help much :(

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/SnapshotAPISpec.scala [hunk 4]
```diff
@@ -199,6 +199,47 @@ class SnapshotAPISpec extends AnyFreeSpecLike with Matchers with BeforeAndAfterA
...
+          val datasetAfterCreation = getDataset(datasetName, projectName, ownerAuthToken)
+          datasetAfterCreation should not be empty
+
+          // delete workspace
+          Rawls.workspaces.delete(projectName, workspaceName)(owner.makeAuthToken())
+
+          // check that the bq dataset has been deleted
+          val datasetAfterDeletion = getDataset(datasetName, projectName, ownerAuthToken)
+          datasetAfterDeletion should be(None)
```
Comment (davidangb) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/SnapshotAPISpec.scala]: I don't know how long it takes for google to delete the dataset once we've issued the request … might need to wrap this in an `eventually` to prevent it from being flaky. I don't know if that's worth doing now to stave off future flakiness or if we should only add `eventually` if/when we see flakiness

**Review by davidangb — APPROVED** [review]
one comment inline, otherwise looks good!

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/SnapshotAPISpec.scala [hunk 5]
```diff
@@ -199,6 +199,47 @@ class SnapshotAPISpec extends AnyFreeSpecLike with Matchers with BeforeAndAfterA
...
+          val workspaceId = resources.head.getMetadata.getWorkspaceId
+          val datasetName = "deltalayer_forworkspace_" + workspaceId.toString.replace('-', '_')
+          val datasetAfterCreation = getDataset(datasetName, projectName, ownerAuthToken)
+          datasetAfterCreation should not be empty
+
+          // delete workspace
+          Rawls.workspaces.delete(projectName, workspaceName)(owner.makeAuthToken())
+
+          // check that the bq dataset has been deleted
+          val datasetAfterDeletion = getDataset(datasetName, projectName, ownerAuthToken)
```
Comment (ungwudik) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/SnapshotAPISpec.scala:237]: Not part of this PR, but it looks like this is calling `getDataset(datasetName: String): F[Option[Dataset]]` which is deprecated. Just noting here that we should consider upgrading to `getDataset(datasetName: BigQueryDatasetName): F[Option[Dataset]]` at some point

**Review by ungwudik — APPROVED** [review]

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/SnapshotAPISpec.scala [hunk 6]
```diff
@@ -193,12 +193,55 @@ class SnapshotAPISpec extends AnyFreeSpecLike with Matchers with BeforeAndAfterA
...
+          eventually {
+            datasetAfterDeletion should be(None)
+          }
```
Comment (davidangb) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/SnapshotAPISpec.scala:240]: the call to getDataset will also need to be inside the eventually - otherwise, it's just a static value that you'll be rechecking for the life of the eventually!
